test/e2e_node/device_manager_test.go

```diff
@@ -266,6 +272,231 @@ var _ = SIGDescribe("Device Manager  [Serial] [Feature:DeviceManager][NodeFeatur
+		an application pod (requesting devices) to successfully run on the node
+		followed by kubelet restart where device plugin doesn't register and the
+		application pod fails with admission error.
+
+		   Breakdown of the steps implemented as part of this e2e test is as follows:
+		   1. Create sample device plugin with AUTO_REGISTER= false which means it
+		      waits for a client to connect to a unix socket exposed at
+		       `/var/lib/kubelet/device-plugins/registered` path.
+		   2. Trigger plugin registeration by connecting to the abovementioned unix socket.
+		   3. Create a test pod requesting devices exposed by the device plugin.
+		   4. Stop kubelet.
+		   5. Remove pods using CRI to ensure new pods are created after kubelet restart.
+		   6. Restart kubelet.
+		   7. Wait for the sample device plugin pod to be running. In this case,
+		      the registration is not triggered.
+		   8. Ensure that resource capacity/allocatable exported by the device plugin is zero.
+		   9. The test pod should fail with `UnexpectedAdmissionError`
+		   10. Delete the test pod.
+		   11. Delete the sample device plugin pod.
+	*/
+	ginkgo.Context("With sample device plugin", func() {
+		var deviceCount int = 2
+		var devicePluginPod *v1.Pod
+
+		// this test wants to reproduce what happened in https://github.com/kubernetes/kubernetes/issues/109595
+		ginkgo.BeforeEach(func(ctx context.Context) {
+			ginkgo.By("Wait for node to be ready")
+			gomega.Eventually(func() bool {
+				nodes, err := e2enode.TotalReady(ctx, f.ClientSet)
+				framework.ExpectNoError(err)
+				return nodes == 1
+			}, time.Minute, time.Second).Should(gomega.BeTrue())
+
+			ginkgo.By("Scheduling a sample device plugin pod")
+			data, err := e2etestfiles.Read(SampleDevicePluginControlRegistrationDSYAML)
+			if err != nil {
+				framework.Fail(err.Error())
+			}
+			ds := readDaemonSetV1OrDie(data)
+
+			dp := &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: sampleDevicePluginName,
+				},
+				Spec: ds.Spec.Template.Spec,
+			}
+
+			devicePluginPod = e2epod.NewPodClient(f).CreateSync(ctx, dp)
+
+			go func() {
+				// Since autoregistration is disabled for the device plugin (as AUTO_REGISTER=false),
+				// device plugin registration needs to be triggerred manually.
+				// This is done by writing to the Unix socket exposed at:
+				// `/var/lib/kubelet/device-plugins/registered`.
+
+				triggerPath := devicePluginDir + "/registered"
+				conn, err := net.Dial("unix", triggerPath)
+				framework.ExpectNoError(err)
+				defer conn.Close()
+
+				_, err = conn.Write([]byte("registering"))
+				framework.ExpectNoError(err)
+			}()
+
+			ginkgo.By("Waiting for devices to become available on the local node")
+			gomega.Eventually(func() bool {
+				node, ready := getLocalTestNode(ctx, f)
+				return ready && numberOfSampleResources(node) > 0
+			}, 5*time.Minute, framework.Poll).Should(gomega.BeTrue())
```

Comment:
`BeTrue` is not going to be informative. You can use https://pkg.go.dev/github.com/onsi/gomega/gcustom to write a better one for this particular check.

Reply:
Can you elaborate on what you would like to be done here. You want me to write up a custom matcher here?

We come to the conclusion that a node is ready if:

1. node status is ready
2. network is setup ( which is done by checking `NetworkUnavailable` condition is not true) 
3. node is not unschedulable
4. device plugin advertises resources on the node

Point 1-3 are exactly what [getLocatTestNode](https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/util.go#L263-L270) does for us. I know it is not ideal that this is returned as a boolean and it would probably be better to have a custom matcher that allows us to `.Should(BeReady)` but that would mean we are reimplmenting a lot of code and I am not sure if that is the best approach.

Reply:
I see that [getLocatTestNode](https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/util.go#L263-L270) is basically a combination of "get node",  `e2enode.IsNodeReady` and `e2enode.IsNodeSchedulable`.

A good solution would be to:
- get the node in the `Eventually` callback
- rewrite `IsNodeReady` and `IsNodeSchedulable` as gomega matchers
- use `gomega.SatisfyAll` with those

But I understand that this is probably more work than you can afford or want to invest.

Some intermediate solution would be to get the node, then use a matcher based on gcustom which invokes `getLocatTestNode` with a failure message that explains the boolean.

Reply:
You can look at e2epod for inspiration. It was recently rewritten to use such a pattern.